### PR TITLE
Provide context to rustfmt-related errors

### PR DESCRIPTION
**Pull request by ja-cop**
If rustfmt is not installed or otherwise not found in $PATH,
uniffi-bindgen now explains that it's rustfmt that wasn't found, instead
of a generic "file not found" error message. Context is also added for
when rustfmt fails to format the scaffolding.

I'm a newbie at Rust and UniFFI, so any feedback is much appreciated! :slightly_smiling_face:

### CODE REVIEW

**Comment by jplatte** [comment]
(I have some thoughts on this, but currently busy)

File: uniffi_bindgen/src/lib.rs [hunk 1]
```diff
@@ -443,6 +444,26 @@ fn parse_udl(udl_file: &Utf8Path) -> Result<ComponentInterface> {
...
+    match Command::new("rustfmt").arg(path).status() {
+        Err(e) => {
+            let context = if e.kind() == ErrorKind::NotFound {
+                "formatting was requested, but rustfmt was not found"
+            } else {
+                "unknown error when calling rustfmt"
+            };
+            Err(e).context(context)
+        }
+        Ok(status) => {
```
Comment (jplatte) [inline uniffi_bindgen/src/lib.rs:457]: This match arm can instead be `.map_err(|e| { let ctx = match e.kind() { ... }; anyhow!(e).context(ctx) })?`.

File: uniffi_bindgen/src/lib.rs [hunk 2]
```diff
@@ -443,6 +444,26 @@ fn parse_udl(udl_file: &Utf8Path) -> Result<ComponentInterface> {
...
+            if !status.success() {
+                Err(anyhow!("rustmt failed when formatting scaffolding. Note: --no-format can be used to skip formatting"))
```
Comment (jplatte) [inline uniffi_bindgen/src/lib.rs:459]: Returning `Err(anyhow!(...))` has a shorthand macro in anyhow called `bail!`.

**Review by jplatte — APPROVED** [review]
Looks good. I guess you posted this because you didn't hqve rustfmt installed in some setup? I wonder whether [prettyplease](https://github.com/dtolnay/prettyplease) could be a better fit for UniFFI's code formatting needs.

File: uniffi_bindgen/src/lib.rs [hunk 3]
```diff
@@ -443,6 +444,20 @@ fn parse_udl(udl_file: &Utf8Path) -> Result<ComponentInterface> {
     ComponentInterface::from_webidl(&udl).context("Failed to parse UDL")
 }
 
+fn format_code_with_rustfmt(path: &Utf8Path) -> Result<()> {
+    let status = Command::new("rustfmt").arg(path).status().map_err(|e| {
+        let ctx = match e.kind() {
+            ErrorKind::NotFound => "formatting was requested, but rustfmt was not found",
+            _ => "unknown error when calling rustfmt",
+        };
+        anyhow!(e).context(ctx)
```
Comment (bendk) [inline uniffi_bindgen/src/lib.rs:453]: Does this context get printed out when `rustfmt` fails?  It seems like you ignore it below, but I could be missing some anyhow magic.

**Review by bendk — CHANGES_REQUESTED** [review]
Seems like a good idea to me, just had a small question about the error context.

**Comment by bendk** [comment]
My feeling is that formatting is a nice to have, but not at all important.  I would consider just printing a warning if `rustfmt` fails.  In fact, I think that's how the bindings generation works.

File: uniffi_bindgen/src/lib.rs (hunk 3 above)
Comment (jplatte) [reply uniffi_bindgen/src/lib.rs:453]: It's not ignored, it gets bubbled out of the function with `?` on the next line.

**Comment by bendk** [comment]
This seems like a step forward, so let's merge.  I'd like at least one other UniFFI dev to chime in before changing the behavior, so I made #1320 for discussion.

**Review by bendk — APPROVED** [review]

**Pull request merged**